server/src/main/java/org/elasticsearch/indices/recovery/PeerRecoverySourceService.java

```diff
@@ -157,6 +183,22 @@ synchronized void cancel(IndexShard shard, String reason) {
             }
         }
 
+        void awaitEmpty() {
+            assert lifecycle.stoppedOrClosed();
```

Comment:
Can we be sure that no new recoveries may start after this method has returned?

Reply:
The indices services close before this one so it should be fine. I added an assertion about the state of the peer recovery source service before adding a recovery though to make sure :)

Reply:
Similarly to my earlier [comment](https://github.com/elastic/elasticsearch/pull/46178/files#r321236268), which is a general comment about code assertions, it would be nice to have more context here in case this assertion ever fails.